packages/conf-binutils/conf-binutils.master/files/find-binutils.ml.in

```diff
@@ -0,0 +1,228 @@
+      sprintf "%s\"%s\"; " s path) ~init:"[" ^ "]"
+
+let is_none = function
+  | None -> true
+  | _ -> false
+
+let failed_to_find cmd =
+  eprintf "%s not found\n" cmd;
+  exit 1
+
+let dedup main files =
+  fst @@
+  List.fold_left files
+    ~init:([], Digests.add main.dgst Digests.empty)
+    ~f:(fun ((files, digests) as acc) ({dgst} as file) ->
+        if Digests.mem dgst digests
+        then acc
+        else file :: files, Digests.add dgst digests)
+
+let dependencies ~init files =
+  List.fold_left files ~init ~f:(fun depends {path; dgst} ->
+      sprintf " [ %S %S ] " path dgst :: depends)
+
+let write t =
+  let deps, vars =
+    Vars.fold (fun name {main; files} (deps, vars) ->
+        let main, files = match main, files with
+          | None, [] -> failed_to_find name
+          | None, main :: files -> main, files
+          | Some main, files -> main, files in
+        let files = dedup main files in
+        let files = main :: files in
+        let deps = dependencies ~init:deps files in
+        let files = sprintf "%s: %S" name (string_of_files files) in
+        deps, files :: vars) t ([], [])  in
+  let oc = open_out "%{_:name}%.config" in
+  let deps = String.concat "\n" deps in
+  let vars = String.concat "\n" vars in
+  fprintf oc {|
+opam-version: "2.0"
+file-depends: [ %s ]
+variables {
+%s
+}
+|} deps vars;
+  close_out oc
+
+let of_env var () =
+  try Some (Sys.getenv var)
+  with Not_found -> None
+
+let of_opam_config var () = match var with
+  | "" -> None
+  | x  -> Some x
+
+let match_files path files test =
+  Array.fold_left files ~init:[] ~f:(fun acc file ->
+      if test file then
+        let path = path / file in
+        {path; dgst = digest path} :: acc
+      else acc)
+
+let update t path  =
+  if Sys.is_directory path then
+    let files = Sys.readdir path in
+    Vars.fold (fun name tool t ->
+        let files = match_files path files tool.test in
+        Vars.add name {tool with files = tool.files @ files} t) t t
+  else t
+
+let collect t pathes = List.fold_left pathes ~f:update ~init:t
+
+let search t path =
+  let rec loop t path =
+    if Sys.is_directory path
+    then
+      Sys.readdir path |>
+      Array.fold_left ~init:t ~f:(fun t file -> loop t @@ path / file)
+    else
+      Vars.fold (fun name tool t ->
+          if tool.test (Filename.basename path)
+          then
+            let file = {path; dgst = digest path} in
+            Vars.add name {tool with files = tool.files @ [file]} t
+          else t) t t in
+  loop t path
+
+let rec first_success = function
+  | [] -> None
+  | f :: fs ->  match f () with
+    | None -> first_success fs
+    | r -> r
+
+let is_dir p = Sys.file_exists p && Sys.is_directory p
+
+let all_pathes () =
+  match of_env "PATH" () with
+  | None -> []
+  | Some r -> List.filter ~f:is_dir (split r ~on:':')
+
+let has_prefix str pref =
+  let len = String.length pref in
+  len <= String.length str &&
+  String.(sub str 0 len = pref)
```

Comment:
it is better to avoid local opens, as if they will add `str`, `len`, or `pref` element to the `String` module, the code will break. 

```suggestion
  String.sub str 0 len = pref
```